### PR TITLE
FIX-#6558: Normalize the number of partitions after `.read_parquet()`

### LINKED ISSUE #6558: Make the number of column partitions dependent on the number of row groups at `.read_parquet()`
At the current implementation, modin's parquet reader tries to create as many column partitions as possible (even if all of them will consist of only 1 column), not considering the amount of row partitions being generated naturally by parquet's row groups.
https://github.com/modin-project/modin/blob/abe20a57b82ae366b2394171d1a7dd50ca1b9cb9/modin/core/io/column_stores/column_store_dispatcher.py#L146

This leads to that `.read_parquet()` naturally produces square-frames (#5296) that performs poorly in modin.
```python
import tempfile
import pandas
import modin.pandas as pd

import modin.config as cfg

cfg.NPartitions.put(16)

NROWS = 100
NCOLS = 16

with tempfile.TemporaryDirectory() as file:
    [
        pandas.DataFrame({f"col{j}": range(NROWS) for j in range(NCOLS)}).to_parquet(
            f"{file}/{i}.parquet"
        )
        for i in range(16)
    ]

    df = pd.read_parquet(file)
    print(df._query_compiler._modin_frame._partitions.shape)  # (16, 16)
```

We may want to change the logic of generating column partitions so if there are already enough row parts then it will only generate column partitions in accordance with the `cfg.MinPartitionSize` parameter, and not in 1 column per 1 partition style.

### PR DESCRIPTION
**Pull request by dchigarev**
## What do these changes do?

This PR brings the following changes to how parquet reading is being distributed:
1. <b>The number of column partitions is now depended on the number of row partitions.</b> Before, the implementation tended to create as much column partitions as possible (it often resulted into that column partitions consisted of only one column) which worked relatively fine if there were no row-groups in the parquet file (and so no row partitions):
    ```python
    NUM_CPUS=16
    parquet_file # has 1 row group and 16 columns
    pd.read_parquet(parquet_file)._partitions.shape # (1, 16) - 1 row partition and 16 column parts
    ```
    However, if there were enough row groups to keep all the workers busy, this excessive amount of column partitions resulted into an over partitioned frame:
    ```python
    file1 # has 9 row groups and 16 columns
    pd.read_parquet(file1)._partitions.shape # (9, 16)
    
    file2 # has 24 row groups and 16 columns
    pd.read_parquet(file2)._partitions.shape # (16, 16) - square partitioned frame
    ```
    Not only this logic generates much more reading kernels than the cores user potentially have, thus slowing down the reading part, but this also tends to generate over partitioned frames that further will slow other operations in the workflow (#5394).

    This logic was changed, so now we first determine how many row parts the output dataframe will have and then depending on the number of remaining partitions (`NPartitions.get() / num_row_parts`) create that number of column partitions. BUT, if the number of columns is greater than the `cfg.MinPartitionSize` parameter, then the columns splitting logic is the same as in [`.from_pandas()`](https://github.com/modin-project/modin/blob/aa752565fad37b0259785d639a0a5bf317f55c24/modin/core/dataframe/pandas/partitioning/partition_manager.py#L788-L789), allowing to create more column partitions (up to the square partitioning).

    Here are few examples of how the new splitting logic for `.read_parquet()` works:
    ```python
    NUM_CPUS = 16
    MIN_PARTITION_SIZE = 32
    
    parquet file schema -> partitioning of modin df
    (row_grps=1, columns=9) -> (row_parts=1, col_parts=9) # no row parts, create col parts as much as possible
    (row_grps=1, columns=18) -> (row_parts=1, col_parts=16) # no row parts, create col parts as much as possible
    (row_grps=9, columns=9) -> (row_parts=9, col_parts=2) # 9 row parts, so only 2 col parts
    (row_grps=9, columns=64) -> (row_parts=9, col_parts=2) # 9 row parts, so only 2 col parts according to MIN_PARTITION_SIZE param
    (row_grps=9, columns=65) -> (row_parts=9, col_parts=3) # 9 row parts, so only 3 col parts according to MIN_PARTITION_SIZE param
    (row_grps=100, columns=9) -> (row_parts=16, col_parts=1) # 16 row parts, so only 1 col part
    (row_grps=100, columns=32) -> (row_parts=16, col_parts=1) # 16 row parts, so only 1 col part according to MIN_PARTITION_SIZE param
    (row_grps=100, columns=1_000) -> (row_parts=16, col_parts=16) # 16 row parts and 16 col parts according to MIN_PARTITION_SIZE param
    ```

2. **More distributed reading across the row groups.** Before, if the number of row groups was greater than `NPartitions` but didn't divide without reminder (18 row groups and 16 `NPartitions`) then the resulted dataframe will have LESS than 16 row partitions:
    ```python
    step = ceil(num_row_groups / npartitions) # ceil(18 / 16) = ceil(1.125) = 2
    row_parts = [
        row_groups[i * step : i * (step + 1)] for i in range(num_row_groups / step)
    ] # len(row_parts) == 9
    ```
    It was changed to allow unequal row partition sizes in order to fill all the row partitions:
    ```python
    step = num_row_groups // npartitions # 18 // 16 = 1
    reminder = num_row_groups % npartitions # 18 % 16 = 2
    row_partition_sizes = [step] * (npartitions - reminder) + [step + 1] * reminder
    # here we will have that 14 workers will read 1 row group and the last 2 workers will read 2 row groups
    ```

### Performance difference

image

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #6558 
- [x] tests <s>added and</s> are passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/core/io/column_stores/column_store_dispatcher.py [hunk 1]
```diff
@@ -163,11 +167,17 @@ def build_columns(cls, columns):
...
-            if columns_length % num_partitions == 0
-            else columns_length // num_partitions + 1
+        if num_row_parts is None:
+            # in column formats we mostly read columns in parallel rather than rows,
+            # so we try to chunk columns as much as possible
+            min_block_size = 1
+        else:
+            num_remaining_parts = round(NPartitions.get() / num_row_parts)
+            min_block_size = min(
+                len(columns) // num_remaining_parts, MinPartitionSize.get()
```
Comment (anmyachev) [inline modin/core/io/column_stores/column_store_dispatcher.py:177]:
```suggestion
                columns_length // num_remaining_parts, MinPartitionSize.get()
```

File: modin/core/io/column_stores/column_store_dispatcher.py [hunk 2]
```diff
@@ -163,11 +167,17 @@ def build_columns(cls, columns):
...
+            )
+        column_splits = compute_chunksize(
+            len(columns), NPartitions.get(), max(1, min_block_size)
```
Comment (anmyachev) [inline modin/core/io/column_stores/column_store_dispatcher.py:180]:
```suggestion
            columns_length, NPartitions.get(), max(1, min_block_size)
```

File: modin/core/io/column_stores/parquet_dispatcher.py [hunk 3]
```diff
@@ -351,19 +355,102 @@ def get_dataset(cls, path, engine, storage_options):
...
+
+        num_splits = min(NPartitions.get(), num_row_groups)
+        parquet_files = dataset.files
+        step = num_row_groups // num_splits
+        # If 'num_splits' does not divide 'num_row_groups' then we can't cover all of
+        # the row groups using the original 'step'. According to the 'reminder'
+        # there has to be that number of partitions that should read 'step + 1'
+        # number of row groups.
+        reminder = num_row_groups % num_splits
+        part_sizes = [step] * (num_splits - reminder) + [step + 1] * reminder
```
Comment (anmyachev) [inline modin/core/io/column_stores/parquet_dispatcher.py:389]:
```suggestion
        parquet_files = dataset.files
        row_groups_per_file = dataset.row_groups_per_file
        num_row_groups = sum(row_groups_per_file)

        if num_row_groups == 0:
            return []

        num_splits = min(NPartitions.get(), num_row_groups)
        part_size = num_row_groups // num_splits
        # If 'num_splits' does not divide 'num_row_groups' then we can't cover all of
        # the row groups using the original 'part_size'. According to the 'reminder'
        # there has to be that number of partitions that should read 'step + 1'
        # number of row groups.
        reminder = num_row_groups % num_splits
        part_sizes = [part_size] * (num_splits - reminder) + [part_size + 1] * reminder
```

**Review by anmyachev — COMMENTED** [review]
LGTM!

**Review by anmyachev — APPROVED** [review]

File: modin/core/io/column_stores/parquet_dispatcher.py [hunk 4]
```diff
@@ -349,19 +353,102 @@ def get_dataset(cls, path, engine, storage_options):
...
+        Parameters
+        ----------
+        dataset : ColumnStoreDataset
+
+        Returns
+        -------
+        list[list[ParquetFileToRead]]
+            Each element in the returned list describes a list of files that a partition has to read.
+        """
+        from modin.core.storage_formats.pandas.parsers import ParquetFileToRead
```
Comment (vnlitvinov) [inline modin/core/io/column_stores/parquet_dispatcher.py:371]: why cannot we import it at the top without `if TYPE_CHECKING` condition?

Comment (dchigarev) [reply modin/core/io/column_stores/parquet_dispatcher.py:371]: it then triggers "an import from a partially initialized module" problem

**Pull request merged**